test/test_testing.py

```diff
@@ -2706,3 +2711,460 @@ def test_stop_stopped_service(self):
         self.assertEqual('foo', foo_info.name)
         self.assertEqual(pebble.ServiceStartup.ENABLED, foo_info.startup)
         self.assertEqual(pebble.ServiceStatus.ACTIVE, foo_info.current)
+
+
+class _PebbleStorageAPIsTestMixin:
+
+    # Override this in classes using this mixin.
+    # This should be set to any non-empty path, but without a trailing /.
+    prefix = None
+
+    def test_push_and_pull_bytes(self):
+        self._test_push_and_pull_data(
+            original_data=b"\x00\x01\x02\x03\x04",
+            encoding=None,
+            stream_class=io.BytesIO)
+
+    def test_push_and_pull_non_utf8_data(self):
+        self._test_push_and_pull_data(
+            original_data='日本語',  # "Japanese" in Japanese
+            encoding='sjis',
+            stream_class=io.StringIO)
+
+    def _test_push_and_pull_data(self, original_data, encoding, stream_class):
+        client = self.client
+        client.push(self.prefix + '/test', original_data, encoding=encoding)
+        with client.pull(self.prefix + '/test', encoding=encoding) as infile:
+            received_data = infile.read()
+        self.assertEqual(original_data, received_data)
+
+        # We also support file-like objects as input, so let's test that case as well.
+        small_file = stream_class(original_data)
+        client.push(self.prefix + '/test', small_file, encoding=encoding)
+        with client.pull(self.prefix + '/test', encoding=encoding) as infile:
+            received_data = infile.read()
+        self.assertEqual(original_data, received_data)
+
+    def test_push_to_non_existant_subdir(self):
+        data = 'data'
+        client = self.client
+
+        with self.assertRaises(pebble.PathError) as cm:
+            client.push(self.prefix + '/nonexistant_dir/test', data, make_dirs=False)
+        self.assertEqual(cm.exception.args[0], 'not-found')
```

Comment:
Would it be better to use `cm.exception.kind` instead of `cm.exception.args[0]` for this and other similar cases?

Reply:
Thanks; I missed that.